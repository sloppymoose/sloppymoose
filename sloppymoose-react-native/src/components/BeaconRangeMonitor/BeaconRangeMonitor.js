import { arrayOf, func, number, shape, string } from 'prop-types'
import Beacons from 'react-native-beacons-manager'
import { DeviceEventEmitter } from 'react-native'
import emptyAry from 'empty/array'
import emptyFn from 'empty/function'
import { filter, intersectionBy, uniqBy, without, xorBy } from 'lodash'
import React from 'react'

export default class BeaconRangeMonitor extends React.Component {
  static defaultProps = {
    children: emptyFn,
    emitter: DeviceEventEmitter,
    regions: emptyAry
  }
  static propTypes = {
    children: func,
    emitter: shape({
      addListener: func
    }),
    regions: arrayOf(
      shape({
        identifier: string.isRequired,
        major: number,
        minor: number,
        uuid: string.isRequired
      }).isRequired
    )
  }
  state = {
    beaconsInRange: emptyAry,
    regionsInRange: emptyAry
  }
  componentWillMount () {
    this.startRanging(this.props.regions)
    this._listener = this.props.emitter.addListener(
      'beaconsDidRange',
      this.handleInRange
    )
  }
  componentWillReceiveProps (nextProps) {
    const { regions } = this.props
    const { regions: nextRegions } = nextProps
    const changedRegions = xorBy(regions, nextRegions, 'uuid')
    if (changedRegions.length) {
      this.stopRanging(intersectionBy(changedRegions, regions, 'uuid'))
      this.startRanging(intersectionBy(changedRegions, nextRegions, 'uuid'))
    }
  }
  componentWillUnmount () {
    this.stopRanging(this.props.regions)
    this._listener.remove()
  }
  handleInRange = ({ region, beacons }) => {
    beacons = without(beacons, ...filter(beacons, { proximity: 'unknown' }))
    if (beacons.length) {
      this.addInRange(beacons, region)
    } else {
      this.removeOutOfRange(region)
    }
  }
  addInRange = (newInRangeBeacons, newInRangeRegion) => {
    let { beaconsInRange, regionsInRange } = this.state
    const oldRegionBeacons = filter(beaconsInRange, {
      uuid: newInRangeRegion.uuid
    })
    beaconsInRange = without(beaconsInRange, ...oldRegionBeacons)
    beaconsInRange = [...newInRangeBeacons, ...beaconsInRange]
    regionsInRange = uniqBy([...regionsInRange, newInRangeRegion], 'uuid')
    this.setState({ beaconsInRange, regionsInRange })
  }
  removeOutOfRange = region => {
    let { beaconsInRange, regionsInRange } = this.state
    const outOfRangeBeacons = filter(beaconsInRange, { uuid: region.uuid })
    beaconsInRange = without(beaconsInRange, ...outOfRangeBeacons)
    regionsInRange = without(regionsInRange, region)
    this.setState({ beaconsInRange, regionsInRange })
  }
  startRanging = regions => {
    regions.map(region => Beacons.startRangingBeaconsInRegion(region))
  }
  stopRanging = regions => {
    regions.map(region => Beacons.stopRangingBeaconsInRegion(region))
  }
  render () {
    return this.props.children(this.state)
  }
}
