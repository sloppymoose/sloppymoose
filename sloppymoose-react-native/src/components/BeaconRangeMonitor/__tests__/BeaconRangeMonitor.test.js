import Beacons from 'react-native-beacons-manager'
import BeaconRangeMonitor from '../BeaconRangeMonitor'
import EventEmitter from 'events'
import {
  FAR,
  IMMEDIATE_1D,
  IMMEDIATE_2D,
  NEAR,
  UNKNOWN
} from './__fixtures__/ranges'
import React from 'react'
import { REGION_A, REGION_B, REGION_C } from './__fixtures__/regions'
import renderer from 'react-test-renderer'

class MockDeviceEmitter extends EventEmitter {}

jest.mock('react-native-beacons-manager', () => ({
  startRangingBeaconsInRegion: jest.fn(),
  stopRangingBeaconsInRegion: jest.fn()
}))

describe('BeaconRangeMonitor', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  describe('#componentWillMount', () => {
    it('starts ranging the specified beacons', done => {
      renderer.create(
        <BeaconRangeMonitor regions={[REGION_A]}>
          {() => {
            expect(
              Beacons.startRangingBeaconsInRegion.mock.calls
            ).toMatchSnapshot()
            done()
            return <i />
          }}
        </BeaconRangeMonitor>
      )
    })
  })

  describe('#componentWillReceiveProps', () => {
    describe('with a different set of regions to range', () => {
      it('stops ranging the deprecated regions', () => {
        const subject = renderer.create(
          <BeaconRangeMonitor regions={[REGION_A, REGION_B]}>
            {() => <i />}
          </BeaconRangeMonitor>
        )
        const instance = subject.getInstance()
        instance.componentWillReceiveProps({ regions: [REGION_B, REGION_C] })
        expect(Beacons.stopRangingBeaconsInRegion.mock.calls).toMatchSnapshot()
      })
      it('starts ranging the newly added regions', () => {
        const subject = renderer.create(
          <BeaconRangeMonitor regions={[REGION_A, REGION_B]}>
            {() => <i />}
          </BeaconRangeMonitor>
        )
        const instance = subject.getInstance()
        instance.componentWillReceiveProps({ regions: [REGION_B, REGION_C] })
        expect(
          Beacons.startRangingBeaconsInRegion.mock.calls.slice(2)
        ).toMatchSnapshot()
      })
      it('does not modify the ranging of legacy regions', () => {
        const subject = renderer.create(
          <BeaconRangeMonitor regions={[REGION_A, REGION_B]}>
            {() => <i />}
          </BeaconRangeMonitor>
        )
        const instance = subject.getInstance()
        instance.componentWillReceiveProps({ regions: [REGION_B, REGION_C] })
        expect(Beacons.startRangingBeaconsInRegion.mock.calls).toMatchSnapshot()
      })
    })

    describe('with no change in monitored regions', () => {
      it('does not stop or restart ranging of existing regions', () => {
        const subject = renderer.create(
          <BeaconRangeMonitor regions={[REGION_A, REGION_B]}>
            {() => <i />}
          </BeaconRangeMonitor>
        )
        const instance = subject.getInstance()
        instance.componentWillReceiveProps({ regions: [REGION_A, REGION_B] })
        expect(Beacons.startRangingBeaconsInRegion.mock.calls).toMatchSnapshot()
        expect(Beacons.stopRangingBeaconsInRegion.mock.calls).toMatchSnapshot()
      })
    })
  })

  describe('when a known beacon is in-range', () => {
    it('adds the region to the list of in-range regions', done => {
      const emitter = new MockDeviceEmitter()
      let renderCount = 0
      renderer.create(
        <BeaconRangeMonitor emitter={emitter} regions={[REGION_A]}>
          {({ beaconsInRange, regionsInRange }) => {
            renderCount = renderCount + 1
            if (renderCount === 2) {
              expect(beaconsInRange).toEqual(NEAR.beacons)
              expect(regionsInRange).toEqual([REGION_A])
              done()
            }
            return <i />
          }}
        </BeaconRangeMonitor>
      ) // initial render count:1
      emitter.emit('beaconsDidRange', NEAR) // render count:2
    })
  })

  describe('when an unknown beacon is in range', () => {
    it('does not add the region to the list of in-range regions', done => {
      const emitter = new MockDeviceEmitter()
      let renderCount = 0
      renderer.create(
        <BeaconRangeMonitor emitter={emitter} regions={[REGION_A]}>
          {({ beaconsInRange, regionsInRange }) => {
            renderCount = renderCount + 1
            if (renderCount === 3) {
              expect(beaconsInRange).toEqual([])
              expect(regionsInRange).toEqual([])
              done()
            }
            return <i />
          }}
        </BeaconRangeMonitor>
      ) // initial render count:1
      emitter.emit('beaconsDidRange', NEAR) // render count:2
      emitter.emit('beaconsDidRange', UNKNOWN) // render count:3
    })
  })

  describe('when an in-range beacon ranges again', () => {
    it('does not duplicate the in-range region in', done => {
      const emitter = new MockDeviceEmitter()
      let renderCount = 0
      renderer.create(
        <BeaconRangeMonitor emitter={emitter} regions={[REGION_A]}>
          {({ beaconsInRange, regionsInRange }) => {
            renderCount = renderCount + 1
            if (renderCount === 3) {
              expect(beaconsInRange).toEqual(NEAR.beacons)
              expect(regionsInRange).toEqual([REGION_A])
              done()
            }
            return <i />
          }}
        </BeaconRangeMonitor>
      ) // initial render count:1
      emitter.emit('beaconsDidRange', NEAR) // render count:2
      emitter.emit('beaconsDidRange', NEAR) // render count:3
    })
  })

  describe('when an in-range beacon moves out of range', () => {
    it('removes the beacon from the list of in-range beacons', done => {
      const emitter = new MockDeviceEmitter()
      let renderCount = 0
      renderer.create(
        <BeaconRangeMonitor emitter={emitter} regions={[REGION_A]}>
          {({ beaconsInRange, regionsInRange }) => {
            renderCount = renderCount + 1
            if (renderCount === 3) {
              expect(beaconsInRange).toEqual([])
              done()
            }
            return <i />
          }}
        </BeaconRangeMonitor>
      ) // initial render count:1
      emitter.emit('beaconsDidRange', NEAR) // render count:2
      emitter.emit('beaconsDidRange', UNKNOWN) // render count:3
    })
    it('removes the region from the list of in-range regions', done => {
      const emitter = new MockDeviceEmitter()
      let renderCount = 0
      renderer.create(
        <BeaconRangeMonitor emitter={emitter} regions={[REGION_A]}>
          {({ beaconsInRange, regionsInRange }) => {
            renderCount = renderCount + 1
            if (renderCount === 3) {
              expect(regionsInRange).toEqual([])
              done()
            }
            return <i />
          }}
        </BeaconRangeMonitor>
      ) // initial render count:1
      emitter.emit('beaconsDidRange', NEAR) // render count:2
      emitter.emit('beaconsDidRange', UNKNOWN) // render count:3
    })
  })

  describe('when one of two in-range beacons moves out of range', () => {
    it('removes the out-of-range beacon from the list of in-range beacons', done => {
      const emitter = new MockDeviceEmitter()
      let renderCount = 0
      renderer.create(
        <BeaconRangeMonitor emitter={emitter} regions={[REGION_A]}>
          {({ beaconsInRange, regionsInRange }) => {
            renderCount = renderCount + 1
            if (renderCount === 3) {
              expect(beaconsInRange).toMatchSnapshot()
              done()
            }
            return <i />
          }}
        </BeaconRangeMonitor>
      ) // initial render count:1
      emitter.emit('beaconsDidRange', IMMEDIATE_2D) // render count:2
      emitter.emit('beaconsDidRange', IMMEDIATE_1D) // render count:3
    })
    it('keeps the region in the list of in-range regions', done => {
      const emitter = new MockDeviceEmitter()
      let renderCount = 0
      renderer.create(
        <BeaconRangeMonitor emitter={emitter} regions={[REGION_A]}>
          {({ beaconsInRange, regionsInRange }) => {
            renderCount = renderCount + 1
            if (renderCount === 3) {
              expect(regionsInRange).toMatchSnapshot()
              done()
            }
            return <i />
          }}
        </BeaconRangeMonitor>
      ) // initial render count:1
      emitter.emit('beaconsDidRange', IMMEDIATE_2D) // render count:2
      emitter.emit('beaconsDidRange', IMMEDIATE_1D) // render count:3
    })
  })

  describe('when an in-range beacon ranges changes proximity', () => {
    it('reflects the change in proximity for the in-range beacon', done => {
      const emitter = new MockDeviceEmitter()
      let renderCount = 0
      renderer.create(
        <BeaconRangeMonitor emitter={emitter} regions={[REGION_A]}>
          {({ beaconsInRange, regionsInRange }) => {
            renderCount = renderCount + 1
            if (renderCount === 3) {
              expect(beaconsInRange[0].proximity).toBe('far')
              done()
            }
            return <i />
          }}
        </BeaconRangeMonitor>
      ) // initial render count:1
      emitter.emit('beaconsDidRange', NEAR) // render count:2
      emitter.emit('beaconsDidRange', FAR) // render count:3
    })
  })

  describe('#componentWillUnmount', () => {
    it('stops ranging the specified beacons', () => {
      const subject = renderer.create(
        <BeaconRangeMonitor regions={[REGION_A]}>
          {() => <i />}
        </BeaconRangeMonitor>
      )
      subject.unmount()
      expect(Beacons.stopRangingBeaconsInRegion.mock.calls).toMatchSnapshot()
    })
  })
})
