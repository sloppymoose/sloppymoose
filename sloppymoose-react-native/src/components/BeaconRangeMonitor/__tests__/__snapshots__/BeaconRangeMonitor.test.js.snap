// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeaconRangeMonitor #componentWillMount starts ranging the specified beacons 1`] = `
Array [
  Array [
    Object {
      "identifier": "REGION A",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_A_1_1",
    },
  ],
]
`;

exports[`BeaconRangeMonitor #componentWillReceiveProps with a different set of regions to range does not modify the ranging of legacy regions 1`] = `
Array [
  Array [
    Object {
      "identifier": "REGION A",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_A_1_1",
    },
  ],
  Array [
    Object {
      "identifier": "REGION B",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_B_1_1",
    },
  ],
  Array [
    Object {
      "identifier": "REGION C",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_C_1_1",
    },
  ],
]
`;

exports[`BeaconRangeMonitor #componentWillReceiveProps with a different set of regions to range starts ranging the newly added regions 1`] = `
Array [
  Array [
    Object {
      "identifier": "REGION C",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_C_1_1",
    },
  ],
]
`;

exports[`BeaconRangeMonitor #componentWillReceiveProps with a different set of regions to range stops ranging the deprecated regions 1`] = `
Array [
  Array [
    Object {
      "identifier": "REGION A",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_A_1_1",
    },
  ],
]
`;

exports[`BeaconRangeMonitor #componentWillReceiveProps with no change in monitored regions does not stop or restart ranging of existing regions 1`] = `
Array [
  Array [
    Object {
      "identifier": "REGION A",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_A_1_1",
    },
  ],
  Array [
    Object {
      "identifier": "REGION B",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_B_1_1",
    },
  ],
]
`;

exports[`BeaconRangeMonitor #componentWillReceiveProps with no change in monitored regions does not stop or restart ranging of existing regions 2`] = `Array []`;

exports[`BeaconRangeMonitor #componentWillUnmount stops ranging the specified beacons 1`] = `
Array [
  Array [
    Object {
      "identifier": "REGION A",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_A_1_1",
    },
  ],
]
`;

exports[`BeaconRangeMonitor when one of two in-range beacons moves out of range keeps the region in the list of in-range regions 1`] = `
Array [
  Object {
    "identifier": "REGION A",
    "major": 1,
    "minor": 1,
    "uuid": "REGION_A_1_1",
  },
]
`;

exports[`BeaconRangeMonitor when one of two in-range beacons moves out of range removes the out-of-range beacon from the list of in-range beacons 1`] = `
Array [
  Object {
    "accuracy": 1,
    "major": 1,
    "minor": 1,
    "proximity": "immediate",
    "rssi": 0,
    "uuid": "REGION_A_1_1",
  },
]
`;
